### PR TITLE
Add post "Test Your Controller at Scale with KWOK"

### PR DESCRIPTION
**Pull request by wzshiming**
This PR adds a blog post "Test Your Controller at Scale with KWOK"

### CODE REVIEW

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+slug: test-your-controller-at-scale-with-kwok
+summary: |
+  Are you a Kubernetes developer looking for a reliable way to test your controllers at scale?
+  Look no further! KWOK makes it easy.
+---
+
+Are you a Kubernetes developer looking for a reliable way to test your controllers at scale?
+Look no further! [KWOK] makes it easy.
+
+In this article, we'll show you how to test your controllers with KWOK, such as Istio and Argo.
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:13]:
```suggestion
This article will demonstrate how to use KWOK to test your controllers, such as Istio and Argo.
```

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+## Testing Istio with KWOK
+
+Latest information can be found at [KWOK with Istio].
+
+Istio is popular component for managing service meshes.
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:19]:
```suggestion
Istio is a popular component for managing service meshes.
```

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+### Deploy Istio
+
+``` bash
+istioctl install -y
+```
+
+### Migrate Controllers to Real Node
+
+The [KWOK] itself cannot actually run a Pod, so we need to bring in [KIND] to run the controller.
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:42]:
```suggestion
The [KWOK] itself cannot actually run a Pod, so we need to use [KIND] to run the controller.
```

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+### Deploy Argo
+
+``` bash
+kubectl create namespace argo
+kubectl apply -n argo -f https://github.com/argoproj/argo-workflows/releases/download/v3.4.8/install.yaml
+```
+
+### Migrate Controllers to Real Node
+
+The [KWOK] itself cannot actually run a Pod, so we need to bring in [KIND] to run the controller.
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:96]:
```suggestion
The [KWOK] itself cannot actually run a Pod, so we need to use [KIND] to run the controller.
```

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+kubectl patch deploy workflow-controller -n argo --type=json -p='[{"op":"add","path":"/spec/template/spec/nodeName","value":"kwok-kwok-control-plane"}]'
+```
+
+### Test Workflow
+
+``` bash
+argo submit -n argo --watch https://raw.githubusercontent.com/argoproj/argo-workflows/master/examples/hello-world.yaml
+```
+
+You will see that the Pod is running on a fake node, it has finished running
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:109]:
```suggestion
You will see that the Pod is running on a fake node and has completed its operation.
```

File: content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+## Getting Involved
+
+If you're interested in participating in future discussions or development related to KWOK, there are several ways to get involved:
+
+- Slack: [#kwok] for general usage discussion, [#kwok-dev] for development discussion. (visit [slack.k8s.io] for a workspace invitation)
+- Open Issues/PRs/Discussions in [sigs.k8s.io/kwok]
+
+We welcome feedback and contributions from anyone who wants to join us in this exciting project.
```
Comment (windsonsea) [inline content/en/blog/_posts/2023-07-01-test-your-controller-at-scale-with-kwok.md:118]:
```suggestion
We welcome feedback and contributions from anyone interested in joining us on this exciting project.
```

**Review by sftim — COMMENTED** [review]
Thanks @wzshiming

We should try to be clear on:
- how this article is different from https://kubernetes.io/blog/2023/03/01/introducing-kwok/, from March this year.
- the “at scale” aspect that the title implies - where do we cover that?

**Comment by wzshiming** [comment]
Thanks, @sftim 

> how this article is different from https://kubernetes.io/blog/2023/03/01/introducing-kwok/, from March this year.

This post is a kind of `Tutorials and walkthroughs`, see these rules https://kubernetes.io/docs/contribute/new-content/blogs-case-studies/#submit-a-post

This is a very typical problem when testing controllers with KWOK, and I have given two examples of cases that I have been asked about.
- https://github.com/kubernetes-sigs/kwok/discussions/582#discussioncomment-6169894
- https://github.com/kubernetes-sigs/kwok/issues/564#issuecomment-1554255052


> the “at scale” aspect that the title implies - where do we cover that?

The main focus is on how to set up a test controller environment using the KWOK, specific test methods vary from controller to controller and are therefore not presented in the content.

The only advantage of using KWOK over other methods is that its scale is no longer limited by hardware conditions, and this may be the only way to test the controller at scale on a laptop.

**Review by nate-double-u — COMMENTED** [review]
Hi @wzshiming,

Thanks for writing this up; however, given that this PR has become a bit stale, and that it covers content already covered in a previous blog post, I'm going to close this.

/close

**Pull request closed without merging**